File: interviewbook/stackandqueue/recursiveReverseStack/stack.java
package interviewbook.stackandqueue.recursiveReverseStack;
import java.util.*;
// this alg using a small trick - recursion stack, very interesting
class stack {
    // try to get bottom element for each recursion in stack
    public int getLast(Stack<Integer> stack) {
        int last = stack.pop();
        if (stack.isEmpty()) {
            return last;
        }
        int ret = getLast(stack);
        stack.push(last);
        return ret;
    }
    // using getLast func to get bottom element, and push reversely
    public void reverse(Stack<Integer> stack) {
        if (stack.isEmpty()) {
            return;
        }
        int last = getLast(stack);
        reverse(stack);
        stack.push(last);
    }

    public static void main(String[] args) {
        stack s = new stack();
        Stack<Integer> st = new Stack<>();
        for (int i = 1; i <=5; i++) {
            st.push(i);
        }
        s.reverse(st);
        String str = "";
        while (!st.isEmpty()) {
            str += st.pop() + "->";
        }
        System.out.println(str + "null");
    }
}

File: interviewbook/stackandqueue/recursiveReverseStack/stack.go
package main

import "fmt"

// Stack primitive
type Stack []int

func (s Stack) isEmpty() bool {
	return len(s) <= 0
}

func (s *Stack) push(val int) {
	*s = append(*s, val)
}

func (s *Stack) pop() int {
	stack := *s
	if stack.isEmpty() {
		return -1
	}
	*s = stack[:len(stack)-1]
	return stack[len(stack)-1]
}

func (s *Stack) peek() int {
	stack := *s
	if stack.isEmpty() {
		return -1
	}
	return stack[len(stack)-1]
}

func getBot(s *Stack) int {
	val := s.pop()
	if s.isEmpty() {
		return val
	}
	ret := getBot(s)
	s.push(val)
	return ret
}

func reverse(s *Stack) {
	if s.isEmpty() {
		return
	}
	val := getBot(s)
	reverse(s)
	s.push(val)
}

func main() {
	var stack Stack
	for i := 1; i <= 5; i++ {
		stack.push(i)
	}
	fmt.Println(stack)
	reverse(&stack)
	fmt.Println(stack)
}
